realsense2_camera/src/realsense_node_factory.cpp

```diff
@@ -326,44 +343,90 @@ void RealSenseNodeFactory::onInit()
 void RealSenseNodeFactory::StartDevice()
 {
 	if (_realSenseNode) _realSenseNode.reset();
-	ros::NodeHandle nh = getNodeHandle();
-	ros::NodeHandle privateNh = getPrivateNodeHandle();
-	// TODO
-	std::string pid_str(_device.get_info(RS2_CAMERA_INFO_PRODUCT_ID));
-	uint16_t pid = std::stoi(pid_str, 0, 16);
-	switch(pid)
+	try
 	{
-	case SR300_PID:
-	case SR300v2_PID:
-	case RS400_PID:
-	case RS405_PID:
-	case RS410_PID:
-	case RS460_PID:
-	case RS415_PID:
-	case RS420_PID:
-	case RS420_MM_PID:
-	case RS430_PID:
-	case RS430_MM_PID:
-	case RS430_MM_RGB_PID:
-	case RS435_RGB_PID:
-	case RS435i_RGB_PID:
-	case RS455_PID:
-	case RS465_PID:
-	case RS_USB2_PID:
-	case RS_L515_PID_PRE_PRQ:
-	case RS_L515_PID:
-		_realSenseNode = std::unique_ptr<BaseRealSenseNode>(new BaseRealSenseNode(nh, privateNh, _device, _serial_no));
-		break;
-	case RS_T265_PID:
-		_realSenseNode = std::unique_ptr<T265RealsenseNode>(new T265RealsenseNode(nh, privateNh, _device, _serial_no));
-		break;
-	default:
-		ROS_FATAL_STREAM("Unsupported device!" << " Product ID: 0x" << pid_str);
-		ros::shutdown();
-		exit(1);
+		ros::NodeHandle nh = getNodeHandle();
+		ros::NodeHandle privateNh = getPrivateNodeHandle();
+		// TODO
+		std::string pid_str(_device.get_info(RS2_CAMERA_INFO_PRODUCT_ID));
+		uint16_t pid = std::stoi(pid_str, 0, 16);
+		switch(pid)
+		{
+		case SR300_PID:
+		case SR300v2_PID:
+		case RS400_PID:
+		case RS405_PID:
+		case RS410_PID:
+		case RS460_PID:
+		case RS415_PID:
+		case RS420_PID:
+		case RS420_MM_PID:
+		case RS430_PID:
+		case RS430_MM_PID:
+		case RS430_MM_RGB_PID:
+		case RS435_RGB_PID:
+		case RS435i_RGB_PID:
+		case RS455_PID:
+		case RS465_PID:
+		case RS_USB2_PID:
+		case RS_L515_PID_PRE_PRQ:
+		case RS_L515_PID:
+			_realSenseNode = std::shared_ptr<BaseRealSenseNode>(new BaseRealSenseNode(nh, privateNh, _device, _serial_no));
+			break;
+		case RS_T265_PID:
+			_realSenseNode = std::shared_ptr<T265RealsenseNode>(new T265RealsenseNode(nh, privateNh, _device, _serial_no));
+			break;
+		default:
+			ROS_FATAL_STREAM("Unsupported device!" << " Product ID: 0x" << pid_str);
+			ros::shutdown();
+			exit(1);
+		}
+		assert(_realSenseNode);
+		_realSenseNode->publishTopics();
+		_initialized = true;
 	}
-	assert(_realSenseNode);
-	_realSenseNode->publishTopics();
+	catch (const rs2::error& e)
+	{
+		ROS_ERROR_STREAM("Exception: " << e.what());
+	}
+}
+
+bool RealSenseNodeFactory::reset()
+{
+	if (!_initialized)
```

Comment:
I could not find the case where the variable "_initialized" is needed. Can you describe such a case?
The test here is redundant, I think:
	// if (!_initialized)
	// {
	// 	return false;
	// }

	// _initialized = false;
The rest of

Reply:
I think I was intending this to prevent multiple overlapping calls to reset if it were to get called twice in quick succession.  But that would only be an issue if the node handler was multi-threaded, which it's not.  So I think you are right that using `if(_device)` is cleaner.